File: src/main/java/pl/yalgrin/playnite/simplesync/service/ChangeService.java
package pl.yalgrin.playnite.simplesync.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.vavr.control.Try;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections4.CollectionUtils;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import pl.yalgrin.playnite.simplesync.domain.objects.Game;
import pl.yalgrin.playnite.simplesync.dto.ChangeDTO;
import pl.yalgrin.playnite.simplesync.dto.GameChangeRequestDTO;
import pl.yalgrin.playnite.simplesync.dto.objects.AbstractObjectDTO;
import pl.yalgrin.playnite.simplesync.dto.objects.GameDTO;
import pl.yalgrin.playnite.simplesync.enums.ObjectType;
import pl.yalgrin.playnite.simplesync.mapper.ChangeMapper;
import pl.yalgrin.playnite.simplesync.repository.ChangeRepository;
import pl.yalgrin.playnite.simplesync.repository.objects.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.*;
import java.util.function.Function;

@Service
@Slf4j
public class ChangeService {
    private final ChangeRepository repository;
    private final ChangeMapper mapper;
    private final ObjectMapper objectMapper;

    private final GameRepository gameRepository;
    private final Map<ObjectType, ObjectRepository<?>> relatedObjectRepositories;

    public ChangeService(ChangeRepository repository, ChangeMapper mapper, ObjectMapper objectMapper,
                         CategoryRepository categoryRepository, GenreRepository genreRepository,
                         PlatformRepository platformRepository, CompanyRepository companyRepository,
                         FeatureRepository featureRepository, TagRepository tagRepository,
                         SeriesRepository seriesRepository, AgeRatingRepository ageRatingRepository,
                         RegionRepository regionRepository, SourceRepository sourceRepository,
                         CompletionStatusRepository completionStatusRepository,
                         FilterPresetRepository filterPresetRepository, GameRepository gameRepository) {
        this.repository = repository;
        this.mapper = mapper;
        this.objectMapper = objectMapper;
        this.gameRepository = gameRepository;
        Map<ObjectType, ObjectRepository<?>> repositoryMap = new LinkedHashMap<>();
        repositoryMap.put(ObjectType.Category, categoryRepository);
        repositoryMap.put(ObjectType.Genre, genreRepository);
        repositoryMap.put(ObjectType.Platform, platformRepository);
        repositoryMap.put(ObjectType.Company, companyRepository);
        repositoryMap.put(ObjectType.Feature, featureRepository);
        repositoryMap.put(ObjectType.Tag, tagRepository);
        repositoryMap.put(ObjectType.Series, seriesRepository);
        repositoryMap.put(ObjectType.AgeRating, ageRatingRepository);
        repositoryMap.put(ObjectType.Region, regionRepository);
        repositoryMap.put(ObjectType.Source, sourceRepository);
        repositoryMap.put(ObjectType.CompletionStatus, completionStatusRepository);
        repositoryMap.put(ObjectType.FilterPreset, filterPresetRepository);
        this.relatedObjectRepositories = repositoryMap;
    }

    @Transactional(readOnly = true)
    public Flux<ChangeDTO> findFromLastId(Long lastId) {
        return repository.findFromLastId(lastId).map(mapper::toDTO);
    }

    public Flux<ChangeDTO> generateChangesForAllObjects() {
        return findMaxId()
                .flatMapMany(maxId ->
                        Flux.mergeSequential(
                                Flux.fromIterable(relatedObjectRepositories.entrySet())
                                        .flatMapSequential(
                                                entry -> findChangesForObjectType(entry.getValue().findAllIds(), maxId,
                                                        entry.getKey())),
                                findChangesForObjectType(gameRepository.findAllIds(), maxId, ObjectType.Game))
                );
    }

    private Mono<Long> findMaxId() {
        return repository.findMaxId().switchIfEmpty(Mono.justOrEmpty(0L));
    }

    private Flux<ChangeDTO> findChangesForObjectType(Flux<Long> ids, Long maxId, ObjectType type) {
        return ids
                .map(id -> ChangeDTO.builder()
                        .id(maxId)
                        .type(type)
                        .objectId(id)
                        .build()
                );
    }

    public Mono<ChangeDTO> saveChange(ChangeDTO changeDTO) {
        return Mono.justOrEmpty(changeDTO)
                .map(mapper::toEntity)
                .flatMap(repository::save)
                .map(mapper::toDTO);
    }

    public Flux<ChangeDTO> generateChangesForGames(GameChangeRequestDTO dto) {
        return Mono.justOrEmpty(dto)
                .filter(d -> CollectionUtils.isNotEmpty(d.getIds()) || CollectionUtils.isNotEmpty(d.getGameIds()))
                .flatMapMany(d -> {
                    CollectedIds collectedIds = new CollectedIds();
                    return fetchGames(d)
                            .doOnNext(game -> {
                                extractObjectUuids(game, collectedIds);
                                collectedIds.getGameIds().add(game.getId());
                            }).then(findIdsForUuids(collectedIds))
                            .thenMany(Flux.mergeSequential(
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getCategoryIds())
                                            .sort(), null, ObjectType.Category),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getGenreIds())
                                            .sort(), null, ObjectType.Genre),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getPlatformIds())
                                            .sort(), null, ObjectType.Platform),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getCompanyIds())
                                            .sort(), null, ObjectType.Company),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getFeatureIds())
                                            .sort(), null, ObjectType.Feature),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getTagIds())
                                            .sort(), null, ObjectType.Tag),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getSeriesIds())
                                            .sort(), null, ObjectType.Series),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getAgeRatingIds())
                                            .sort(), null, ObjectType.AgeRating),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getRegionIds())
                                            .sort(), null, ObjectType.Region),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getSourceIds())
                                            .sort(), null, ObjectType.Source),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getCompletionStatusIds())
                                            .sort(), null, ObjectType.CompletionStatus),
                                    findChangesForObjectType(Flux.fromIterable(collectedIds.getGameIds())
                                            .sort(), null, ObjectType.Game)
                            ));
                });
    }

    private void extractObjectUuids(Game game, CollectedIds collectedIds) {
        GameDTO targetDto = Try.of(
                        () -> objectMapper.readValue(game.getSavedData().asArray(), GameDTO.class))
                .getOrElse(GameDTO::new);
        if (targetDto.getCategories() != null) {
            targetDto.getCategories().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getCategoryUuids()::add);
        }
        if (targetDto.getGenres() != null) {
            targetDto.getGenres().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getGenreUuids()::add);
        }
        if (targetDto.getPlatforms() != null) {
            targetDto.getPlatforms().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getPlatformUuids()::add);
        }
        if (targetDto.getDevelopers() != null) {
            targetDto.getDevelopers().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getCompanyUuids()::add);
        }
        if (targetDto.getPublishers() != null) {
            targetDto.getPublishers().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getCompanyUuids()::add);
        }
        if (targetDto.getFeatures() != null) {
            targetDto.getFeatures().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getFeatureUuids()::add);
        }
        if (targetDto.getTags() != null) {
            targetDto.getTags().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getTagUuids()::add);
        }
        if (targetDto.getSeries() != null) {
            targetDto.getSeries().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getSeriesUuids()::add);
        }
        if (targetDto.getAgeRatings() != null) {
            targetDto.getAgeRatings().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getAgeRatingUuids()::add);
        }
        if (targetDto.getRegions() != null) {
            targetDto.getRegions().stream().map(AbstractObjectDTO::getId)
                    .forEach(collectedIds.getRegionUuids()::add);
        }
        Optional.ofNullable(targetDto.getSource())
                .map(AbstractObjectDTO::getId)
                .ifPresent(id -> collectedIds.getSourceUuids().add(id));
        Optional.ofNullable(targetDto.getCompletionStatus())
                .map(AbstractObjectDTO::getId)
                .ifPresent(id -> collectedIds.getCompletionStatusUuids().add(id));
    }

    private Mono<Void> findIdsForUuids(CollectedIds collectedIds) {
        return Mono.when(findCategories(collectedIds),
                findGenres(collectedIds),
                findPlatforms(collectedIds),
                findCompanies(collectedIds),
                findFeatures(collectedIds),
                findTags(collectedIds),
                findSeries(collectedIds),
                findAgeRatings(collectedIds),
                findRegions(collectedIds),
                findSources(collectedIds),
                findCompletionStatuses(collectedIds));
    }

    private Flux<Game> fetchGames(GameChangeRequestDTO dto) {
        Set<String> alreadyFetchedIds = Collections.synchronizedSet(new HashSet<>());
        return getByIdPairs(dto, alreadyFetchedIds)
                .flatMapMany(list -> Flux.merge(Flux.fromIterable(list), getByIds(dto, alreadyFetchedIds)));
    }

    private Mono<List<Game>> getByIdPairs(GameChangeRequestDTO dto, Set<String> alreadyFetchedIds) {
        if (dto == null || dto.getGameIds() == null) {
            return Mono.just(new ArrayList<>());
        }
        return Flux.fromIterable(dto.getGameIds())
                .filter(i -> i != null && i.getGameId() != null && i.getPluginId() != null)
                .flatMap(i -> gameRepository.findByGameIdAndPluginId(i.getGameId(), i.getPluginId()))
                .doOnNext(g -> alreadyFetchedIds.add(g.getPlayniteId()))
                .collectList().switchIfEmpty(Mono.just(new ArrayList<>()));
    }

    private Flux<Game> getByIds(GameChangeRequestDTO dto, Set<String> alreadyFetchedIds) {
        if (dto == null || dto.getIds() == null) {
            return Flux.empty();
        }
        return Flux.fromIterable(dto.getIds()).filter(i -> !alreadyFetchedIds.contains(i)).buffer(100)
                .flatMap(gameRepository::findByPlayniteIdIn);
    }

    private Mono<Void> findCategories(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getCategoryUuids, CollectedIds::getCategoryIds,
                ObjectType.Category);
    }

    private Mono<Void> findGenres(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getGenreUuids, CollectedIds::getGenreIds,
                ObjectType.Genre);
    }

    private Mono<Void> findPlatforms(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getPlatformUuids, CollectedIds::getPlatformIds,
                ObjectType.Platform);
    }

    private Mono<Void> findCompanies(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getCompanyUuids, CollectedIds::getCompanyIds,
                ObjectType.Company);
    }

    private Mono<Void> findFeatures(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getFeatureUuids, CollectedIds::getFeatureIds,
                ObjectType.Feature);
    }

    private Mono<Void> findTags(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getTagUuids, CollectedIds::getTagIds,
                ObjectType.Tag);
    }

    private Mono<Void> findSeries(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getSeriesUuids, CollectedIds::getSeriesIds,
                ObjectType.Series);
    }

    private Mono<Void> findAgeRatings(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getAgeRatingUuids, CollectedIds::getAgeRatingIds,
                ObjectType.AgeRating);
    }

    private Mono<Void> findRegions(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getRegionUuids, CollectedIds::getRegionIds,
                ObjectType.Region);
    }

    private Mono<Void> findSources(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getSourceUuids, CollectedIds::getSourceIds,
                ObjectType.Source);
    }

    private Mono<Void> findCompletionStatuses(CollectedIds collectedIds) {
        return findIdsForUuidsForObject(collectedIds, CollectedIds::getCompletionStatusUuids,
                CollectedIds::getCompletionStatusIds, ObjectType.CompletionStatus);
    }

    private Mono<Void> findIdsForUuidsForObject(CollectedIds collectedIds,
                                                Function<CollectedIds, Set<String>> uuidExtractor,
                                                Function<CollectedIds, Set<Long>> idExtractor, ObjectType objectType) {
        return Mono.just(collectedIds)
                .map(uuidExtractor)
                .flatMapMany(Flux::fromIterable)
                .buffer(100)
                .flatMap(ids -> relatedObjectRepositories.get(objectType).findIdsByPlayniteIdIn(ids))
                .doOnNext(id -> idExtractor.apply(collectedIds).add(id)).then();
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    private static class CollectedIds {
        private Set<String> categoryUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> categoryIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> genreUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> genreIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> platformUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> platformIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> companyUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> companyIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> featureUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> featureIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> tagUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> tagIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> seriesUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> seriesIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> ageRatingUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> ageRatingIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> regionUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> regionIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> sourceUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> sourceIds = Collections.synchronizedSet(new HashSet<>());
        private Set<String> completionStatusUuids = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> completionStatusIds = Collections.synchronizedSet(new HashSet<>());
        private Set<Long> gameIds = Collections.synchronizedSet(new HashSet<>());
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/change/service/ChangeService.kt
package pl.yalgrin.playnite.simplesync.change.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import pl.yalgrin.playnite.simplesync.change.dto.ChangeDTO
import pl.yalgrin.playnite.simplesync.change.dto.GameChangeRequestDTO
import pl.yalgrin.playnite.simplesync.change.mapper.ChangeMapper
import pl.yalgrin.playnite.simplesync.change.mapper.ChangeMessageMapper
import pl.yalgrin.playnite.simplesync.change.repository.ChangeRepository
import pl.yalgrin.playnite.simplesync.client.message.ChangeMessage
import pl.yalgrin.playnite.simplesync.common.enums.ObjectType
import pl.yalgrin.playnite.simplesync.library.domain.Game
import pl.yalgrin.playnite.simplesync.library.dto.GameDTO
import pl.yalgrin.playnite.simplesync.library.repository.*
import pl.yalgrin.playnite.simplesync.security.getSessionClientId
import pl.yalgrin.playnite.simplesync.util.asObject
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import reactor.kotlin.core.publisher.toFlux
import reactor.kotlin.core.publisher.toMono
import java.util.concurrent.ConcurrentHashMap

@Service
class ChangeService(
    val repository: ChangeRepository,
    val changeMessageMapper: ChangeMessageMapper,
    val mapper: ChangeMapper,
    categoryRepository: CategoryRepository,
    genreRepository: GenreRepository,
    platformRepository: PlatformRepository,
    companyRepository: CompanyRepository,
    featureRepository: FeatureRepository,
    tagRepository: TagRepository,
    seriesRepository: SeriesRepository,
    ageRatingRepository: AgeRatingRepository,
    regionRepository: RegionRepository,
    sourceRepository: SourceRepository,
    completionStatusRepository: CompletionStatusRepository,
    filterPresetRepository: FilterPresetRepository,
    val gameRepository: GameRepository
) {
    private val relatedObjectRepositories: Map<ObjectType, ObjectRepository<*>>

    init {

        val repositoryMap: MutableMap<ObjectType, ObjectRepository<*>> = LinkedHashMap()
        repositoryMap[ObjectType.CATEGORY] = categoryRepository
        repositoryMap[ObjectType.GENRE] = genreRepository
        repositoryMap[ObjectType.PLATFORM] = platformRepository
        repositoryMap[ObjectType.COMPANY] = companyRepository
        repositoryMap[ObjectType.FEATURE] = featureRepository
        repositoryMap[ObjectType.TAG] = tagRepository
        repositoryMap[ObjectType.SERIES] = seriesRepository
        repositoryMap[ObjectType.AGE_RATING] = ageRatingRepository
        repositoryMap[ObjectType.REGION] = regionRepository
        repositoryMap[ObjectType.SOURCE] = sourceRepository
        repositoryMap[ObjectType.COMPLETION_STATUS] = completionStatusRepository
        repositoryMap[ObjectType.FILTER_PRESET] = filterPresetRepository
        this.relatedObjectRepositories = repositoryMap
    }

    @Transactional(readOnly = true)
    fun findFromLastId(lastId: Long?): Flux<ChangeMessage> {
        return getSessionClientId()
            .flatMapMany { repository.findFromLastId(lastId, it) }
            .map { changeMessageMapper.toMessage(it) }
    }

    @Transactional(readOnly = true)
    fun generateChangesForAllObjects(): Flux<ChangeMessage> {
        return findMaxId()
            .flatMapMany { maxId ->
                Flux.mergeSequential(
                    Flux.fromIterable(relatedObjectRepositories.entries)
                        .flatMapSequential { entry ->
                            entry.value.findAllIds().collectList().flatMapMany {
                                findChangesForObjectType(
                                    it,
                                    maxId,
                                    entry.key
                                )
                            }
                        },
                    gameRepository.findAllIds().collectList().flatMapMany {
                        findChangesForObjectType(it, maxId, ObjectType.GAME)
                    }
                )
            }
    }

    private fun findMaxId(): Mono<Long> {
        return repository.findMaxId().defaultIfEmpty(0L)
    }

    @Transactional(readOnly = true)
    fun generateChangesForGames(dto: GameChangeRequestDTO): Flux<ChangeMessage> {
        return dto.toMono()
            .filter { !it.ids.isNullOrEmpty() || !it.gameIds.isNullOrEmpty() }
            .flatMapMany { d ->
                val collectedIds = CollectedIds()
                fetchGames(d)
                    .flatMap { g ->
                        g.savedData.asObject(GameDTO::class.java)
                            .doOnNext { extractObjectUuids(it, collectedIds) }
                            .thenReturn(g)
                    }
                    .doOnNext { game ->
                        game.id?.let { collectedIds.getIds(ObjectType.GAME).add(it) }
                    }.then(findIdsForUuids(collectedIds))
                    .thenMany(
                        Flux.fromIterable(relatedObjectRepositories.keys)
                            .collectList()
                            .map { it.plus(ObjectType.GAME) }
                            .flatMapMany { Flux.fromIterable(it) }
                            .flatMapSequential { objectType ->
                                Mono.fromCallable { collectedIds.getIds(objectType) }
                                    .subscribeOn(Schedulers.boundedElastic())
                                    .flatMapMany { ids -> findChangesForObjectType(ids, null, objectType) }
                            }
                    )

            }
    }

    private fun fetchGames(dto: GameChangeRequestDTO): Flux<Game> {
        val alreadyFetchedIds = ConcurrentHashMap.newKeySet<String>()
        return getByIdPairs(dto, alreadyFetchedIds)
            .flatMapMany { list ->
                Flux.merge(
                    Flux.fromIterable(list),
                    getByIds(dto, alreadyFetchedIds)
                )
            }
    }

    private fun getByIdPairs(
        dto: GameChangeRequestDTO,
        alreadyFetchedIds: MutableSet<String>
    ): Mono<List<Game>> {
        if (dto.gameIds.isNullOrEmpty()) {
            return Mono.just(emptyList())
        }
        //TODO !!
        return Flux.fromIterable(dto.gameIds)
            .filter { i -> i?.gameId != null && i.pluginId != null }
            .flatMap { i ->
                gameRepository.findByGameIdAndPluginId(
                    i.gameId,
                    i.pluginId
                )
            }
            .doOnNext { g -> alreadyFetchedIds.add(g.playniteId!!) }
            .collectList()
            .defaultIfEmpty(emptyList<Game>())
    }

    private fun getByIds(dto: GameChangeRequestDTO, alreadyFetchedIds: Set<String>): Flux<Game> {
        if (dto.ids == null) {
            return Flux.empty()
        }
        return Flux.fromIterable(dto.ids)
            .filter { i -> !alreadyFetchedIds.contains(i) }
            .buffer(100)
            .flatMap { playniteId -> gameRepository.findByPlayniteIdIn(playniteId) }
    }

    private fun extractObjectUuids(targetDto: GameDTO, collectedIds: CollectedIds) {
        targetDto.categories?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.CATEGORY).add(it) } }
        }
        targetDto.genres?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.GENRE).add(it) } }
        }
        targetDto.platforms?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.PLATFORM).add(it) } }
        }
        targetDto.publishers?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.COMPANY).add(it) } }
        }
        targetDto.developers?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.COMPANY).add(it) } }
        }
        targetDto.features?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.FEATURE).add(it) } }
        }
        targetDto.tags?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.TAG).add(it) } }
        }
        targetDto.series?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.SERIES).add(it) } }
        }
        targetDto.ageRatings?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.AGE_RATING).add(it) } }
        }
        targetDto.regions?.let { l ->
            l.stream().map { obj -> obj.id }
                .forEach { e -> e?.let { collectedIds.getUuids(ObjectType.REGION).add(it) } }
        }
        targetDto.source?.id?.let { collectedIds.getUuids(ObjectType.SOURCE).add(it) }
        targetDto.completionStatus?.id?.let { collectedIds.getUuids(ObjectType.COMPLETION_STATUS).add(it) }
    }

    private fun findIdsForUuids(collectedIds: CollectedIds): Mono<Void> {
        return Flux.fromIterable(relatedObjectRepositories.keys)
            .flatMap { objectType ->
                Mono.fromCallable {
                    collectedIds.getUuids(objectType)
                }.subscribeOn(Schedulers.boundedElastic())
                    .flatMapMany { Flux.fromIterable(it) }
                    .buffer(100)
                    .flatMap { relatedObjectRepositories[objectType]?.findIdsByPlayniteIdIn(it) ?: Flux.empty() }
                    .doOnNext { id -> collectedIds.getIds(objectType).add(id) }
            }
            .then()
    }

    private fun findChangesForObjectType(ids: Collection<Long>, maxId: Long?, type: ObjectType): Flux<ChangeMessage> {
        return ids.toFlux().map { id ->
            ChangeMessage(
                id = maxId,
                type = type,
                objectId = id
            )
        }
    }

    fun saveChange(changeDTO: ChangeDTO): Mono<ChangeDTO> {
        return changeDTO.toMono()
            .map { dto -> mapper.toEntity(dto) }
            .flatMap { entity -> repository.save(entity) }
            .map { entity -> mapper.toDTO(entity) }
    }
}

private data class CollectedIds(
    val objectUuids: MutableMap<ObjectType, MutableSet<String>> = ConcurrentHashMap<ObjectType, MutableSet<String>>(),
    val objectIds: MutableMap<ObjectType, MutableSet<Long>> = ConcurrentHashMap<ObjectType, MutableSet<Long>>()
) {
    fun getUuids(type: ObjectType): MutableSet<String> = objectUuids.getOrPut(type) { ConcurrentHashMap.newKeySet() }

    fun getIds(type: ObjectType): MutableSet<Long> = objectIds.getOrPut(type) { ConcurrentHashMap.newKeySet() }
}